File: rosalind/majority-element.go
/*

Problem
An array A[1..n] is said to have a majority element if more than half of its entries are the same.

Given: A positive integer k≤20, a positive integer n≤10^4, and k arrays of size n containing positive integers not exceeding 10^5.

Return: For each array, output an element of this array occurring strictly more than n/2 times if such element exists, and "-1" otherwise.

Source: Algorithms by Dasgupta, Papadimitriou, Vazirani. McGraw-Hill. 2006.

Sample Dataset
4 8
5 5 5 5 5 5 5 5
8 7 7 7 1 7 3 7
7 1 6 5 10 100 1000 1
5 1 6 7 1 1 10 1

Sample Output
5 7 -1 -1

Discussion
It is not difficult to develop a divide-and-conquer algorithm checking whether a given array of size n contains a majority element in O(nlogn) time.
It is interesting to note that there is also a linear time algorithm and it is also based on divide-and-conquer.

*/

package main

func main() {
	assert(majority([]int{5, 5, 5, 5, 5, 5, 5, 5}) == 5)
	assert(majority([]int{8, 7, 7, 7, 1, 7, 3, 7}) == 7)
	assert(majority([]int{7, 1, 6, 5, 10, 100, 1000, 1}) == -1)
	assert(majority([]int{5, 1, 6, 7, 1, 1, 10, 1}) == -1)
	assert(majority([]int{}) == -1)
}

func assert(x bool) {
	if !x {
		panic("assertion failed")
	}
}

// https://en.wikipedia.org/wiki/Boyer%E2%80%93Moore_majority_vote_algorithm
func majority(a []int) int {
	n, m, c := len(a), 0, 1
	for i := 1; i < n; i++ {
		if a[m] == a[i] {
			c++
		} else {
			c--
		}
		if c == 0 {
			m, c = i, 1
		}
	}

	c = 0
	for i := range a {
		if a[m] != a[i] {
			continue
		}
		if c++; c > n/2 {
			return a[m]
		}
	}
	return -1
}

File: leetcode/majority-element.c
/*

Given an array of size n, find the majority element. The majority element is the element that appears more than ⌊ n/2 ⌋ times.

You may assume that the array is non-empty and the majority element always exist in the array.

*/

#include <assert.h>
#include <stdio.h>

#define nelem(x) (sizeof(x) / sizeof(x[0]))

// boyer-moore majority voting algorithm
int
majority(int *a, size_t n)
{
	size_t i, m, c;

	if (n == 0)
		return -1;

	m = 0;
	c = 1;
	for (i = 1; i < n; i++) {
		(a[m] == a[i]) ? c++ : c--;
		if (c == 0) {
			m = i;
			c = 1;
		}
	}
	return a[m];
}

int
main(void)
{
	int a[] = {3, 2, 3};
	int b[] = {2, 2, 1, 1, 1, 2, 2};

	assert(majority(a, nelem(a)) == 3);
	assert(majority(b, nelem(b)) == 2);

	return 0;
}
